### PR TITLE
Add `Copy` button to gists

### PR DESCRIPTION
**Pull request by deadcoder0904**
Everything is working fine. There is an issue while copying Markdown file. I know it can't be done by using the same method so I tried sending XHR request which gives some Github error. So if u have any approach for copying content for Markdown file, suggest me. And tell me if there is something to be changed :smiley:

### CODE REVIEW

**Comment by deadcoder0904** [comment]
@sindresorhus Can u merge this?

**Comment by sindresorhus** [comment]
> Everything is working fine. There is an issue while copying Markdown file. I know it can't be done by using the same method so I tried sending XHR request which gives some Github error.

I think we should just not show the Copy button if it's not displaying source code, like a rendered markdown file. That would be the simplest solution.

**Comment by sindresorhus** [comment]
There needs to be some spacing between the Copy and Raw buttons. Use the same spacing as the other buttons.

File: extension/copy-gist.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+'use strict';
+
+window.addGistCopyButton = () => {
+	// Button already added (partial page nav), or non-text file
+	if ($('.copy-btn').length > 0) {
+		return;
+	}
+	const $gistsSibling = $('div.file-actions>a.btn.btn-sm');
+	for (let i = 0; i < $gistsSibling.length; i++) {
```
Comment (sindresorhus) [inline extension/copy-gist.js]: Use a for-of loop.

File: extension/copy-gist.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+		const gistUri = $gistsSibling[i].href;
+		$(`<a href="${gistUri}" class="btn btn-sm copy-btn">Copy</a>`).insertBefore($gistsSibling[i]);
+	}
+
+	$(document).on('click', '.copy-btn', e => {
+		e.preventDefault();
+		const fileContents = $('.js-file-line-container').get(0).innerText;
```
Comment (sindresorhus) [inline extension/copy-gist.js]: This needs to handle when a gists has multiple files.

Can also use the `.text()` jQuery method.

File: extension/copy-gist.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/* globals utils */
+
+'use strict';
+
+window.addGistCopyButton = () => {
+	// Button already added (partial page nav), or non-text file
+	if ($('.copy-btn').length > 0) {
+		return;
+	}
+	const $gistsSibling = $('div.file-actions>a.btn.btn-sm');
```
Comment (sindresorhus) [inline extension/copy-gist.js]:
```js
const $gistsSibling = $('.file-actions > .btn.btn-sm');
```

**Comment by sindresorhus** [comment]
Needs to be added to the Highlights section in the readme.

**Comment by deadcoder0904** [comment]
On it Sir!! :smile:

**Comment by deadcoder0904** [comment]
Done :+1:

File: extension/copy-gist.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+window.addGistCopyButton = () => {
+	// Button already added (partial page nav), or non-text file
+	if ($('.copy-btn').length > 0) {
+		return;
+	}
+	const $gistsSibling = $('.file-actions > .btn.btn-sm');
+
+	for (const anchor of $gistsSibling) {
+		const gistUri = anchor.href;
+		if (gistUri.split('.').pop().toLowerCase() !== 'md') {
```
Comment (sindresorhus) [inline extension/copy-gist.js]: There are more file types that get rendered than just Markdown, like structured text. That's why I asked for detecting when it's not a source code view.

File: extension/copy-gist.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+			$(`<a href="${gistUri}" class="btn btn-sm copy-btn" id="gists">Copy</a>`).insertBefore(anchor);
```
Comment (sindresorhus) [inline extension/copy-gist.js]: You're now creating lots of `id`'s. `id`'s are supposed to be unique.

File: extension/copy-gist.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+		}
+	}
+
+	$(document).on('click', '.copy-btn', e => {
+		e.preventDefault();
+		const fileContents = $(`#${e.currentTarget.offsetParent.id}`).find('.js-file-line-container')[0].innerText;
```
Comment (sindresorhus) [inline extension/copy-gist.js]: I think a better solution would be to find the closest below source code container. Also use `.text()` like mentioned before.

File: yarn.lock [hunk 7]
```diff
@@ -0,0 +1,3673 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (sindresorhus) [inline yarn.lock]: Don't add this.

File: extension/content.css [hunk 8]
```diff
@@ -532,3 +532,8 @@ td.blob-code.blob-code-deletion:before {
 .diffbar > .diffstat {
 	float: right;
 }
+
+/* copy button for gists margin */
+#gists {
```
Comment (sindresorhus) [inline extension/content.css]: Should be `.copy-btn`

File: extension/copy-gist.js (hunk 4 above)
Comment (deadcoder0904) [inline extension/copy-gist.js]: Didn't get this :slightly_frowning_face: Idk the file extensions which don't have a source code view... I think u mean that there is no numbering on the RHS like in the above `.js` file. Right ??

File: extension/copy-gist.js (hunk 5 above)
Comment (deadcoder0904) [inline extension/copy-gist.js]: Agreed :+1:

File: extension/content.css (hunk 8 above)
Comment (deadcoder0904) [inline extension/content.css]: Yep

File: extension/copy-gist.js (hunk 6 above)
Comment (deadcoder0904) [inline extension/copy-gist.js]: Will definitely try this :smile:

File: extension/copy-gist.js (hunk 4 above)
Comment (sindresorhus) [inline extension/copy-gist.js]: Yes, just find a CSS class for the source code view that is not present in the rendered view.

File: extension/copy-gist.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+window.addGistCopyButton = () => {
+	// Button already added (partial page nav), or non-text file
+	if ($('.copy-btn').length > 0) {
+		return;
+	}
+	const $gistsSibling = $('.file-actions > .btn.btn-sm');
+
+	for (let i = 0; i < $gistsSibling.length; i++) {
+		const sourceCodeClass = $(`#${$gistsSibling[i].offsetParent.id}>div`).eq(1)[0].className;
```
Comment (sindresorhus) [inline extension/copy-gist.js]: This is a bit inefficient and convoluted. Would be better to find the common container for both the buttons and the source code and then find the source code from there.

File: extension/copy-gist.js (hunk 6 above)
Comment (sindresorhus) [inline extension/copy-gist.js]: Ah, ok. Makes sense.

**Comment by deadcoder0904** [comment]
Alrighty done :smiley_cat: I think its complete now :joy:

**Comment by sindresorhus** [comment]
Looks great now. Thanks for working on this :)

**Pull request merged**